### PR TITLE
Track and close intermediate realm references

### PR DESCRIPTION
**Pull request by rorbech**
This makes the `Realm` frozen and tracks all intermediate frozen realm references and closes them when the realm is closed.

### CODE REVIEW

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt [hunk 1]
```diff
@@ -134,9 +140,23 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
     private suspend fun updateRealmPointer(newRealm: NativePointer, newVersion: VersionId) {
         realmPointerMutex.withLock {
             log.debug("Updating Realm version: $version -> $newVersion")
-            if (newVersion >= version) {
-                realmReference = RealmReference(this, newRealm)
+            val newReference = RealmReference(this, newRealm)
+            val intermediateReference = if (newVersion >= version) {
+                val version = Pair(realmReference.dbPointer, WeakReference(realmReference))
+                realmReference = newReference
+                version
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/Realm.kt:147]: You are returning `version` here and a Pair in the else. This seems wrong?

Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt:147]: Also the naming in this section is a bit hard to follow, we have`newReference`, `RealmReference`,  `WeakReference`, `intermediateReference` and `intermediateReferences`, at least it makes my head spin a little when trying to reason about it. Not exactly sure I have a better proposal though.

File: test/build.gradle.kts [hunk 2]
```diff
@@ -48,6 +48,7 @@ kotlin {
                 // Our current compiler plugin tests only runs on JVM, so makes sense to keep them
                 // for now, but ideally they should go to the compiler plugin tests.
                 implementation("io.realm.kotlin:cinterop:${Realm.version}")
+                implementation("org.jetbrains.kotlinx:atomicfu:${Versions.atomicfu}")
```
Comment (cmelchior) [inline test/build.gradle.kts:51]: Does this work? I had problems using atomicfu in the test project

File: test/src/androidMain/kotlin/io/realm/util/PlatformUtils.kt [hunk 3]
```diff
@@ -41,4 +42,36 @@ actual object PlatformUtils {
...
+
+    @Suppress("ExplicitGarbageCollectionCall")
+    actual fun triggerGC() {
+        for (i in 1..2) {
+            allocGarbage(0)
+            SystemClock.sleep(100)
+            System.gc()
+            System.runFinalization()
+        }
+        SystemClock.sleep(5000) // 5 seconds to give the GC some time to process
```
Comment (cmelchior) [inline test/src/androidMain/kotlin/io/realm/util/PlatformUtils.kt:54]: Is there any, even anecdotal, evidence this does anything useful? Sleeping for 5 seconds is pretty heavy.

**Review by cmelchior — CHANGES_REQUESTED** [review]

File: test/build.gradle.kts (hunk 2 above)
Comment (rorbech) [reply test/build.gradle.kts:51]: This is required since I am asserting on the internal `realm.intermediateReferences.value`. So not applying it.

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt (hunk 1 above)
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt]: The name was wrong, so it was actually returning a reference. I tried to separate the _reference tracking_ into a method to hide building the pairs, weak references, etc. Hopefully it is a bit more clear now.

File: test/src/androidMain/kotlin/io/realm/util/PlatformUtils.kt (hunk 3 above)
Comment (rorbech) [reply test/src/androidMain/kotlin/io/realm/util/PlatformUtils.kt:55]: I just moved the code from https://github.com/realm/realm-kotlin/pull/300/files#diff-8243d5deaa5b3325464776989c0be50c922a20b96ac486be0e9fdf60f026c011L165 to `PlatformUtils` to enable using `triggerGC` in common tests. Only comment that I accidentally didn't copy is `Empiric approach to trigger GC`. I added it again, but maybe @nhachicha can elaborate a bit?

**Comment by rorbech** [comment]
@cmelchior Ready for another round. The sample app tests are failling as `Notifications are not available on frozen lists since they do not change.`. I guess it doesn't really make sense to try to bring in order before we more of your pieces in play too. So I would consider to just to merge it if it passes review this time.

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt [hunk 4]
```diff
@@ -131,15 +138,35 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
...
+    // Must only be called with realmPointerMutex locked
+    private fun trackNewAndCloseExpiredReferences(realmReference: RealmReference) {
+        val expiredReferences = intermediateReferences.value.filter { (_, ref) -> ref.get() == null }
+        expiredReferences.map { (pointer, _) ->
+            log.debug("Closing unreferenced version: ${RealmInterop.realm_get_version_id(pointer)}")
+            RealmInterop.realm_close(pointer)
+        }
+        intermediateReferences.value = intermediateReferences.value
+            .plus(Pair(realmReference.dbPointer, WeakReference(realmReference)))
+            .subtract(expiredReferences)
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt]: 👍 Tweaked it a bit to only iterate the list once.

We have to build a new set as the atomic value is frozen on assignment, so we cannot update the existing one on Native.

**Comment by rorbech** [comment]
I merged this even though the sample app test fails. The failure is due to listening for notifications on frozen lists that will be targeted in #296. Added a TODO item in #291 on fixing the sample app.
